### PR TITLE
Don't install packages that could mess packaging up

### PR DESCRIPTION
**Pull request by kennethreitz**

### CODE REVIEW

File: vendor/pip-pop/pip-clean [hunk 1]
```diff
@@ -0,0 +1,40 @@
...
+  -h --help     Show this screen.
+"""
+from docopt import docopt
+
+BAD_PACKAGES = ['appdirs', 'packaging', 'pyparsing', 'six', 'setuptools', 'distribute']
+
+def main():
+    args = docopt(__doc__, version='pip-grep')
+    req_file = args['<req-file>']
+
```
Comment (cclauss) [inline vendor/pip-pop/pip-clean]: Could this be streamlined?

```python
def good_package(line):
	package_name = line.split('=')[0].split('<')[0].split('>')[0]
	return package_name not in BAD_PACKAGES

lines = [line for line in f if good_package(line)]
```

Comment (kennethreitz) [inline vendor/pip-pop/pip-clean]: I want to do as little parsing as possible.

Comment (cclauss) [inline vendor/pip-pop/pip-clean]: Understood...  see other package names that begin with but do not end with `six` for instance https://pypi.python.org/pypi?%3Aaction=search&term=Six&submit=search

File: bin/steps/setuptools [hunk 2]
```diff
@@ -0,0 +1,6 @@
+#!/usr/bin/env bash
+
+# Syntax sugar.
+source $BIN_DIR/utils
+
+pip-clean requirements.txt
```
Comment (hone) [inline bin/steps/setuptools]: Should we have a message that tells people we're replacing stuff in their `requirements.txt` with things in the Python buildpack or overwriting their `requirements.txt` file at all.

Also, how does this affect pipenv?

File: vendor/pip-pop/pip-clean (hunk 1 above)
Comment (kennethreitz) [inline vendor/pip-pop/pip-clean]: I was planning to cross that bridge when I came to it, which I honestly doubt would happen any time soon. I'll see what I can do though.

**Comment by hone** [comment]
Should we have a message that tells people we're replacing stuff in their requirements.txt with things in the Python buildpack or overwriting their requirements.txt file at all.

Also, how does this affect pipenv?

**Comment by kennethreitz** [comment]
Pipenv has this functionality built-in. 

I don't think a message is needed, but I could add one saying "Removing packaging libraries from requirements.txt".

**Comment by kennethreitz** [comment]
In my mind, if someone has any of these listed in their requirements.txt, it is accidental, not intentional.

**Comment by kennethreitz** [comment]
e.g. they don't understand what's going on. so, i don't want to overload them with information.

**Comment by hone** [comment]
@kennethreitz is there a place you could link them to that would help?

**Comment by kennethreitz** [comment]
No.

**Comment by kennethreitz** [comment]
This is a new problem, and one that is temporary and will go away soon (but we need to have this code for people using this version range of setuptools locally for the future). 

It's not well-understood.

File: vendor/pip-pop/pip-clean (hunk 1 above)
Comment (kennethreitz) [inline vendor/pip-pop/pip-clean]: @cclauss i updated the code; integrated what you suggested and made a few improvements

**Comment by kennethreitz** [comment]
Waiting, waiting, waiting... for CI to pass.

File: bin/steps/setuptools (hunk 2 above)
Comment (jayvdb) [inline bin/steps/setuptools]: missing EOL

File: vendor/pip-pop/pip-clean [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+BAD_PACKAGES = ['appdirs', 'packaging', 'pyparsing', 'six', 'setuptools', 'distribute']
+
+
+def good_package(line):
+    package_name = line.split('=')[0].split('<')[0].split('>')[0].split(' ')[0].split('#')[0].split('\n')[0]
+    return package_name not in BAD_PACKAGES
+
+def main():
+    args = docopt(__doc__, version='pip-grep')
```
Comment (jayvdb) [inline vendor/pip-pop/pip-clean]: `version='pip-grep'`

Comment (kennethreitz) [inline vendor/pip-pop/pip-clean]: thanks!

File: bin/steps/setuptools (hunk 2 above)
Comment (kennethreitz) [inline bin/steps/setuptools]: thanks!

**Comment by jayvdb** [comment]
> In my mind, if someone has any of these listed in their requirements.txt, it is accidental, not intentional.
> e.g. they don't understand what's going on. so, i don't want to overload them with information.

The more usual explanation is they have put them in because they have minimum and maximum requirements.
`setuptools` installs `pkg_resources` which is the main mechanism people use for fetching plugins (implicit namespaces is the newer alternative).
And more importantly, because of the fact that pip doesnt have a proper resolver, sometimes it is *necessary* to mention one of these packages in `requirements.txt` so that pip follows a particular path in its resolution strategy.  You remove one of them of `requirements.txt`, and pip behaves differently when resolving indirect dependencies many levels deep.
At the very least, there should be a note to say why the deploy is using different versions from the versions they see when they deploy the app somewhere else.

File: vendor/pip-pop/requirements.txt [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+python-dateutil==2.6.0
+pytz==2017.2
+pytzdata==2017.2
+regex==2017.5.26
+requests==2.17.3
+ruamel.ordereddict==0.4.9
+ruamel.yaml==0.14.12
+tzlocal==1.4
+urllib3==1.21.1
+
```
Comment (jayvdb) [inline vendor/pip-pop/requirements.txt]: missing EOL.

**Comment by kennethreitz** [comment]
Related: https://github.com/heroku/heroku-buildpack-python/issues/393

File: vendor/pip-pop/pip-clean [hunk 5]
```diff
@@ -0,0 +1,33 @@
...
+Options:
+  -h --help     Show this screen.
+"""
+from docopt import docopt
+
+BAD_PACKAGES = ['appdirs', 'packaging', 'pyparsing', 'six', 'setuptools', 'distribute']
+
+
+def good_package(line):
+    package_name = line.split('=')[0].split('<')[0].split('>')[0].split(' ')[0].split('#')[0].split('\n')[0]
```
Comment (kennethreitz) [inline vendor/pip-pop/pip-clean]: '\n' is included in each line

Comment (jayvdb) [inline vendor/pip-pop/pip-clean]: Buggy. c.f. https://github.com/heroku/heroku-buildpack-python/issues/400

**Comment by edmorley** [comment]
> This is a new problem, and one that is temporary and will go away soon

Was this to work around setuptools unvendoring their dependencies in v34.0.0 (which they've then reverted in v36.0.0) and relying on `install_requires` instead? (Which presumably could cause issues if the versions they required conflicted with those in requirements.txt?)

**Comment by kennethreitz** [comment]
@edmorley yes

**Pull request merged**